File: main/java/com/helloworld/youth/util/activityutil/ActivityManager.java
package com.helloworld.youth.util.activityutil;

import android.app.Activity;

import java.util.Stack;

public class ActivityManager {
	// 栈
	private static Stack<Activity> activityStack = new Stack<Activity>();

	private static ActivityManager instance;

	private ActivityManager() {
	}

	// 管理器
	public static ActivityManager getScreenManager() {
		if (instance == null) {
			instance = new ActivityManager();
		}
		return instance;
	}

	// 栈顶弹出
	public void popActivity() {
		Activity activity = activityStack.lastElement();
		if (activity != null) {
			activity.finish();
			activity = null;
		}
	}

	// 指定activity弹出
	public void popActivity(Activity activity) {
		if (activity != null) {
			activity.finish();
			activityStack.remove(activity);
			activity = null;
		}
	}

	// 弹出到指定activity
	public void popAssignActivity(Class clazz) {
		for (int a = activityStack.size() - 1; a >= 0; a--) {
			if (activityStack.get(a).getClass().equals(clazz)) {
				ActivityManager.getScreenManager().popActivity(
						activityStack.get(a));
				return;
			}

		}
	}

	// 返回栈顶activity
	public Activity currentActivity() {
		if (activityStack == null || activityStack.size() == 0) {
			return null;
		}
		Activity activity = activityStack.lastElement();

		return activity;
	}

	// 增加（压栈）
	public void pushActivity(Activity activity) {
		if (activityStack == null) {
			activityStack = new Stack<Activity>();
		}

		activityStack.add(activity);
	}

	// 保留一个指定activity
	public void popAllActivityExceptOne(Class<?> cls) {
		while (true) {
			Activity activity = currentActivity();
			if (activity == null) {
				break;
			}
			if (activity.getClass().equals(cls)) {
				break;
			}
			popActivity(activity);
		}
	}

	// 弹出所有activity
	public void popAllActivity() {
		while (true) {
			Activity activity = currentActivity();
			if (activity == null) {
				break;
			}
			popActivity(activity);
		}
	}

	// 弹出所有n个activity
	public void popActivity(int num) {
		for (int i = 0; i < num; i++) {
			Activity activity = currentActivity();
			if (activity == null) {
				break;
			}
			popActivity(activity);
		}
	}

	public Stack<Activity> getActivityStack() {
		return activityStack;
	}

	public Activity getActivity(String activityName) {
		for (int i = 0; i < activityStack.size(); i++) {
			if (activityStack.get(i).getClass().getName().equals(activityName)) {
				return activityStack.get(i);
			}
		}
		return null;
	}
}

File: main/java/com/helloworld/youth/util/activityutil/ActivityManager.kt
package com.helloworld.util.activityutil

import android.app.Activity
import java.util.*

class ActivityManagerKT private constructor() {
    // 栈顶弹出
    fun popActivity() {
        var activity = activityStack.lastElement()
        if (activity != null) {
            activity.finish()
        }
    }

    // 指定activity弹出
    fun popActivity(activity: Activity?) {
        if (activity != null) {
            activity.finish()
            activityStack.remove(activity)
        }
    }

    // 弹出到指定activity
    fun popAssignActivity(clazz: Class<*>) {
        for (a in activityStack.indices.reversed()) {
            if (activityStack[a].javaClass == clazz) {
                screenManager!!.popActivity(
                    activityStack[a]
                )
                return
            }
        }
    }

    // 返回栈顶activity
    fun currentActivity(): Activity? {
        return if ( activityStack.size == 0) {
            null
        } else activityStack.lastElement()
    }

    // 增加（压栈）
    fun pushActivity(activity: Activity) {

        activityStack.add(activity)
    }

    // 保留一个指定activity
    fun popAllActivityExceptOne(cls: Class<*>) {
        while (true) {
            val activity = currentActivity() ?: break
            if (activity.javaClass == cls) {
                break
            }
            popActivity(activity)
        }
    }

    // 弹出所有activity
    fun popAllActivity() {
        while (true) {
            val activity = currentActivity() ?: break
            popActivity(activity)
        }
    }

    // 弹出所有n个activity
    fun popActivity(num: Int) {
        for (i in 0 until num) {
            val activity = currentActivity() ?: break
            popActivity(activity)
        }
    }


    fun getActivity(activityName: String): Activity? {
        for (i in activityStack.indices) {
            if (activityStack[i].javaClass.name == activityName) {
                return activityStack[i]
            }
        }
        return null
    }

    companion object {
        // 栈
        private var activityStack: Stack<Activity> = Stack()
        private var instance: ActivityManagerKT? = null
        // 管理器
        val screenManager: ActivityManagerKT?
            get() {
                if (instance == null) {
                    instance = ActivityManagerKT()
                }
                return instance
            }
        fun get(): ActivityManagerKT{
            //细心的小伙伴肯定发现了，这里不用getInstance作为为方法名，是因为在伴生对象声明时，内部已有getInstance方法，所以只能取其他名字
            return instance!!
        }
        fun getStack(): Stack<Activity>{
            //细心的小伙伴肯定发现了，这里不用getInstance作为为方法名，是因为在伴生对象声明时，内部已有getInstance方法，所以只能取其他名字
            return activityStack
        }

    }
}